### DISCUSSION TITLE
Rule of Thumb / best practices for ROHD class and function

### DISCUSSION THREAD

**Discussion by quekyj** [intel/rohd#332] [Q&A] [answered]
In general, when should we create a class or function in dart/ROHD during the design? Any rule of thumb or best practices?

**Comment by mkorbel1** [comment, marked as answer]
From ROHD's perspective, one time when you do want to make a class is when you want to define a `Module` (e.g. `class ___ extends Module`).  When you want something to be a `Module` is sometimes similar to when you want to make something a `module` in SystemVerilog.  For example, a top-level hierarchy or backend synthesis partition should be a `Module`.  A `Module` always translates into a `module` when converted to SystemVerilog.

SystemVerilog uses `module`s and `function`s as common ways to group functionality.  In ROHD, you are not limited to only `Module`s to group reusable functionality, and ROHD will not generate a SystemVerilog `function`.

ROHD, as a generator framework, enables you to write _software_ that _generates hardware_.  You can use any types of software collections, architectures, design patterns, etc. including classes and functions to develop your hardware.  The only time you need to explicitly define a `Module` as a layer of hierarchy is when you actually want there to be a formal module boundary there.  Any logic generated by your software has its hierarchy and placement determined by connectivity alone, so it will end up in the "appropriate" module hierarchy as long as you follow the input/output rules of ROHD (only consume `input`s of a `Module` within that `Module`, and communicate externally through `output`s of that `Module`).

So to summarize, make a `class` that extends `Module` when you want an explicit hardware hierarchy level to be added.  Otherwise, you can do whatever you like (including other classes and functions), since it's just software!

**Reply by quekyj** [reply]
So, it really depend on how the user wants to structure their code.
